src/hypothesis/strategies.py

```diff
@@ -815,20 +814,59 @@ def decimals(
       allow_nan.
     - If both min_value and max_value are not None, it is an error to enable
       allow_infinity.
+    - If places is not None, all values will have the given number of digits
+      after the decimal point.  (places=2 -> 1.23, 4.00, etc.)
+    - If precision is not None, values will be generated with at most the
+      given number of digits, including trailing zeros.
 
     Where not explicitly ruled out by the bounds, all of infinity, -infinity
     and NaN are possible values generated by this strategy.
-
+    Where places and precision are both set, it is an invalid operation to
+    generate a ``value >= 10 ** (precision - places)`` and thus an error if
+    max_value is set and would allow such values.
     """
+    check_valid_bound(min_value, 'min_value')
+    check_valid_bound(max_value, 'max_value')
+    check_valid_interval(min_value, max_value, 'min_value', 'max_value')
+
+    check_valid_integer(places)
+    if places is not None and places < 0:
+        raise InvalidArgument('places=%s must be >= 0' % places)
+    check_valid_integer(precision)
+    if precision is not None and precision <= 0:
+        raise InvalidArgument('precision=%s must be > 0' % precision)
+    check_valid_interval(places, precision, 'places', 'precision')
+
+    if places is not None and precision is not None:
+        if max_value is None:
+            # Cap values at 999/1000 of the validity threshold
+            max_value = 10 ** (precision - places) - 10 ** (-places - 3)
+        elif max_value >= 10 ** (precision - places):
```

Comment:
300, 310, 320, ..., 1000, 1100, ..., etc.  Add trailing zeros to fill out places.

That's the plan anyway, it appears that the C-style signal handling and context management is worse than I expected so mostly it's just throwing exceptions.